add skywire-cli subcommands

0pcom opened the pull request:
Did you run `make format && make check`?
yes

~~Fixes pull request review issue from skycoin/skybian#134 ; creating http endpoint for querying the hypervisor's public key~~
*re-implemented srvpk into skybian much more simply [here](https://github.com/the-skycoin-project/skybian/blob/rework/script/srvpk.service)



 Changes:	
- ~~Creates HTTP endpoint  for `skywire cli visor pk` with `skywire-cli hv srvpk`~~
- `skywire-cli hv` subcommands (see systray PR)
- Minor restructuring of `skywire-cli` visor commands
- fix api endpoints must return errors for remotevisors and vpnserverslist

How to test this PR:

Clone my fork, checkout srvpk branch and start a visor
```
git clone https://github.com/the-skycoin-project/skywire
cd skywire
git checkout srvpk
make run-source
```

~~start the http endpoint in a new terminal or tab~~
~~(`ctrl`+`shift`+`t`)~~
~~```~~
~~go run cmd/skywire-cli/skywire-cli.go hv srvpk~~
~~```~~

~~query the http endpoint at [http://127.0.0.1:7998](http://127.0.0.1:7998) in the browser~~
~~or in a new terminal or tab using curl~~
~~(`ctrl`+`shift`+`t`)~~
~~```~~
~~curl http://127.0.0.1:7998~~
~~```~~

~~The output should be the same as the following:~~
~~```~~
~~go run cmd/skywire-cli/skywire-cli.go visor pk~~
~~```~~
~~without the trailing newline~~

Commit:
add cli commands from other PR and srvpk http endpoint

Commit:
commit selected changes from other branch

Commit:
add 'skywire-cli hv srvpk' subcommand

jdknives commented on cmd/skywire-cli/commands/hv/hvdmsg/hvdmsg.go:
```diff
@@ -0,0 +1,90 @@
+
+func init() {
+	RootCmd.AddCommand(dmsgUICmd)
+	dmsgUICmd.Flags().StringVarP(&path, "input", "i", "", "read from specified config file")
```
I dont see this flag used yet in the commands.

0pcom replied:
![image](https://user-images.githubusercontent.com/36607567/176554631-82c37857-c598-4e0e-8ce6-a35f88ee4cfc.png)

jdknives commented on cmd/skywire-cli/commands/hv/hvdmsg/hvdmsg.go:
```diff
@@ -0,0 +1,90 @@
+				client := rpcClient()
+				overview, err := client.Overview()
+				if err != nil {
+					log.Fatal("Failed to connect:", err)
```
Maybe leave a message here to ask the user if the visor is running...Applies for some other commands as well.

0pcom replied:
noted

0pcom replied:
fixed more than just this instance with next commit

jdknives commented on cmd/skywire-cli/commands/hv/hvdmsg/hvdmsg.go:
```diff
@@ -0,0 +1,90 @@
+				url = fmt.Sprintf("http://127.0.0.1:8000/pty/%s", overview.PubKey.Hex())
+			}
+		} else {
+			url = fmt.Sprintf("http://127.0.0.1:8000/pty/%s", pk)
```
Would the hostname need to be configurable?

0pcom replied:
I think the circumstance for that is rare, but you are correct that the behavior does assume its running locally.

Configurable, Im not sure.

We could print the LAN ip address / have a flag flag that would print that.

jdknives commented on cmd/skywire-cli/commands/hv/hvdmsg/hvdmsg.go:
```diff
@@ -0,0 +1,90 @@
+	},
+}
+
+var dmsgURLCmd = &cobra.Command{
```
This command does not work and does not show up in the help of `skywire-cli hv`.

0pcom replied:
fixed next commit

jdknives commented on cmd/skywire-cli/commands/hv/hvskychat/root.go:
```diff
@@ -0,0 +1,21 @@
+package hvskychat
```
I dont think this will ever be used to be honest. Skychat is a test application that we dont ship with the packages.

0pcom replied:
this was just my attempt to evenly implement this functionality for all the UIs.

we have a community member who has a massive rework of skychat in progress.

Otherwise, yes, I think the usage would be slim.

jdknives commented on cmd/skywire-cli/commands/hv/root.go:
```diff
@@ -0,0 +1,23 @@
+// RootCmd contains commands that interact with the skywire-visor
+var RootCmd = &cobra.Command{
+	Use:   "hv",
+	Short: "open HVUI in browser",
```
Is this actually used? This certainly is not the command that actually opens the hypervisor UI as that is `skywire-cli hv ui`.

jdknives commented on cmd/skywire-cli/commands/hv/srvpk.go:
```diff
@@ -0,0 +1,36 @@
+	Use:   "srvpk",
+	Short: "http endpoint for `skywire-cli visor pk`",
+	Run: func(_ *cobra.Command, _ []string) {
+		if !sourcerun {
```
Whats this flag used for? What is the purpose of having it?

jdknives commented on cmd/skywire-cli/commands/hv/srvpk.go:
```diff
@@ -0,0 +1,36 @@
+			skywirecli = "go run cmd/skywire-cli/skywire-cli.go"
+		}
+		for {
+			_, err := script.Exec(`nc -l -p ` + port + ` -e '` + skywirecli + ` visor pk -w'`).Stdout()
```
Why are we using `netcat` here? Anything we cant do with Go specificallt?

jdknives commented on cmd/skywire-cli/commands/hv/srvpk.go:
```diff
@@ -0,0 +1,36 @@
+			_, err := script.Exec(`nc -l -p ` + port + ` -e '` + skywirecli + ` visor pk -w'`).Stdout()
+			if err != nil {
+				err.Error()
+				//os.Exit(1)
```
What does this comment do here?

jdknives commented on cmd/skywire-cli/commands/visor/info.go:
```diff
@@ -12,14 +13,17 @@ import (
 	RootCmd.AddCommand(hvpkCmd)
 	hvpkCmd.Flags().StringVarP(&path, "input", "i", "", "path of input config file.")
 	hvpkCmd.Flags().BoolVarP(&pkg, "pkg", "p", false, "read from /opt/skywire/skywire.json")
+	hvpkCmd.Flags().BoolVarP(&web, "http", "w", false, "format as http response")
```
Why do we need to format this as http response?

jdknives commented on cmd/skywire-cli/commands/visor/shutdown.go:
```diff
@@ -0,0 +1,22 @@
+	Use:   "halt",
+	Short: "stop a running visor",
+	Run: func(_ *cobra.Command, args []string) {
+		rpcClient().Shutdown() //nolint
```
We are not checking any return value here. We dont know if the service was actually shutdown.

0pcom replied:
Nothing is returned if the visor is shut down!

If it did reply, it wouldnt be shut down.

The only way to really tell is to try it again or try some other rpc call.

It does actually work from my tests, but there is no good way to distinguish between the errors, i think

jdknives commented on cmd/skywire-cli/commands/visor/start.go:
```diff
@@ -0,0 +1,45 @@
+	if usrLvl.Username == "root" {
+		root = true
+	}
+	//enforce root permissions for VPN operability
```
Do we need to enforce that? Should that not be done by the package automatically?

0pcom replied:
The package does not currently use this command directly in any script.

Binaries are installed to /usr/bin not sbin

as far as i knew, it needed to be enforced but i have no mechanism for doing so in the package itself.

this may need to be changed if we can get rootless vpn working

0pcom replied:
changed to use the appropriate config for the level you invoke the command with

* It does not check that the config exists already.

if you use root, it uses the config generated by `skywire-cli config gen -p`

if you run as the user, it uses the config generated by `skywire-cli config gen -u`

jdknives commented on cmd/skywire-cli/commands/visor/start.go:
```diff
@@ -0,0 +1,45 @@
+	}
+}
+
+var startCmd = &cobra.Command{
```
Does not run for me. Just prints out the help if I run `skywire-cli visor start`.

0pcom replied:
the subcommand was only available as root at the time you tested it. I've rewritten to support starting skywire at the user level.

jdknives commented on pkg/visor/api.go:
```diff
@@ -567,6 +573,31 @@ func (v *Visor) GetAppConnectionsSummary(appName string) ([]appserver.Connection
+		SK:       v.conf.SK,
+		DiscAddr: v.conf.Launcher.ServiceDisc,
+	}, &http.Client{Timeout: time.Duration(1) * time.Second}, "")
+	//ctx, _ := context.WithTimeout(context.Background(), 7*time.Second)
```
Left comment?

0pcom replied:
will clean up

0pcom replied:
fixed in next commit

jdknives commented on pkg/visor/api.go:
```diff
@@ -567,6 +573,31 @@ func (v *Visor) GetAppConnectionsSummary(appName string) ([]appserver.Connection
 }
 
+// VPNServers gets available public VPN server from service discovery URL
+func (v *Visor) VPNServers() []string {
```
Why are we not returning an error here? This is supposed to be relayed back to the CLI invoking it I would suppose. So it should be logged there imo.

0pcom replied:
correct, an error should be returned as well. will change.

0pcom replied:
fixed in next commit

Commit:
Merge branch 'develop' into srvpk

Commit:
fix typography; restructure skywire-cli config, visor, and hv subcommands. Remove hv srvpk subcommand. Limit hv skychat command to linux. Allow skywire-cli visor start as non root

Commit:
fix typography; restructure skywire-cli config, visor, and hv subcommands. Remove hv srvpk subcommand. Limit hv skychat command to linux. Allow skywire-cli visor start as non root

Commit:
make changes per pull request review

0pcom commented:
I've fixed a lot of the typography in the skywire-cli subcommands & made capitalization consistent for subcommands

I've also done some minor rearrangements to just the files and folders containing some subcommands; to overall reduce the number of packages being imported & reduce some code.

Ive fixed the errors returned by the vpnservers and remotevisors api endpoints

removed the `hv srvpk` subcommand as I've re-implemented that to skybian.
